### DISCUSSION TITLE
Running more than 12 transactions at once using Promise.all throws error

### DISCUSSION THREAD

**Discussion by arashi-dev** [prisma/prisma#18573] [Q&A] [answered]
Hello there.

I have a function for my api which uses $transaction inside of it.
I am testing it using Vitest, and for simulating requests, I've run the function 20 times using Promise.all, like so:

###### the function
```ts
const verifyEmail = (data) => {
    return prisma.$transaction(async tx => {
       // some reading
       // some updaing
    })
}
```

###### test
```ts
test("run verifyEmail multiple times", async () => {
    await Promise.all(myCredentials.map(data => verifyEmail(data)))
})
```

and the error I get:
```
Transaction API error: Unable to start a transaction in the given time.
```

I won't get error if the `myCrendentials` only have 12 items. but if the items increased to more than 12, prisma throws the above error.
any idea why this happens?

p.s. increasing the timeout and maxWait didn't help

**Comment by ludralph** [comment]
Hi @arashi-dev 👋 

Have you tried the recommendation in this issue [comment](https://github.com/prisma/prisma/issues/10218#issuecomment-1119416894) about using array transactions?

**Reply by arashi-dev** [reply]
hi, thanks for your response.

yes, I did. but my transaction is not very heavy. even if I make a simple update query inside of it, it will throw the error after running more than 12 or 13 transactions at a time.
imagine thousands of users are going to request the API and run the transaction. BOOM :)

**Reply by ludralph** [reply]
Can you share the code so i can reproduce this?

**Reply by arashi-dev** [reply]
I think I found the problem @ludralph 
I am making some queries in prisma middleware.
so it throws this error because somehow the queries inside of middleware are using the main prisma client BUT inside a transaction.

is there any way to execute them using the running transaction client (tx argument)?
or I create some util functions and use them instead of the middleware?

**Reply by ludralph** [reply]
One way to ensure that the middleware uses the transaction client is to pass the transaction client as an argument to the middleware function. You can modify the middleware function to accept the `tx` argument and use it to execute queries within the transaction context. Alternatively, you could create a utility function as you suggested.

**Reply by arashi-dev** [reply]
@ludralph 
I did it using the util function. that works man :)
thanks for your helping

**Comment by arashi-dev** [comment, marked as answer]
Found the issue.
I had some extra queries inside Prisma middleware. and when I was updating my record inside the transaction, the middleware was making the transaction much heavier than it should be.

so, I came up with creating a util function instead and doing the process inside the transaction directly.
now I can run ~1000 transactions at a time :)

**Reply by ludralph** [reply]
Thank you for sharing what worked for you. It's beneficial.
